### PR TITLE
Auto-apply chat template in `SequenceGenerator` and `SequenceGeneratorAdapter`, if available

### PR DESCRIPTION
**Pull request by leloykun**
This PR auto-applies chat templates by default when using instruct/chat models. Doesn't support LlamaCPP for now tho.

---

### Why?

Instruct/Chat models tend to be annoyingly template dependent (i.e. they perform worse if the prompts don't follow the chat template used in the finetuning step). And the more and longer they are finetuned, the worse the problem gets. Hence this PR.

Also see this issue https://github.com/outlines-dev/outlines/issues/987 raised by @lapp0 

---

### Interface changes

This PR has minimal impact on the interface. It just changes the default behavior of the generation step.

However, this feature can be disabled either on the creation of the generator:
```python
generator = outlines.generate.choice(model, ["Positive", "Negative"], apply_chat_template=False)
```
Or when calling the generator
```python
answer = generator(prompt, apply_chat_template=False)
```

### CODE REVIEW

File: outlines/generate/api.py [hunk 1]
```diff
@@ -529,3 +567,22 @@ def stream(
             self.sampling_params,
             **model_specific_params,
         )
+
+
+def apply_chat_template_util(model, prompts: Union[str, List[str]]) -> List[str]:
+    from outlines.models.transformers import TransformerTokenizer
+
+    if isinstance(prompts, str):
+        prompts = [prompts]
```
Comment (lapp0) [inline outlines/generate/api.py:576]: I think the signature should be `List[str]` -> `List[str]` and raise an error if a list isn't passed. In `transformers.py`, this function is called after the prompts are normalized to 2D anyways.

File: outlines/generate/api.py [hunk 2]
```diff
@@ -529,3 +567,22 @@ def stream(
...
+    if not isinstance(model.tokenizer, TransformerTokenizer):
+        warnings.warn(
+            "Chat template is only supported for `Transformer` models for now. The raw prompts will be used instead."
+        )
+        return prompts
+    tokenizer: "TransformerTokenizer" = model.tokenizer
+    if getattr(tokenizer.tokenizer, "chat_template", None) is None:
```
Comment (lapp0) [inline outlines/generate/api.py:583]: Should we use `default_chat_template` instead?

https://huggingface.co/docs/transformers/v4.34.0/chat_templating#how-do-chat-templates-work

File: outlines/generate/api.py [hunk 3]
```diff
@@ -516,9 +549,14 @@ def stream(
         max_tokens: Optional[int] = None,
         stop_at: Optional[Union[str, List[str]]] = None,
         seed: Optional[int] = None,
+        apply_chat_template: Optional[bool] = None,
         **model_specific_params,
     ):
         """Return a text generator from a prompt or a list of prompts."""
+        if apply_chat_template is None:
```
Comment (lapp0) [inline outlines/generate/api.py:556]: To make this pythonic, we should have one obvious way of applying a chat template. IMO the argument should only be accepted in the constructor.

File: outlines/models/transformers.py [hunk 4]
```diff
@@ -100,6 +116,24 @@ def convert_token_to_string(self, token: str) -> str:
...
+        Template for one-turn chat.
+        """
+        return (
+            [self.SYSTEM_MESSAGE.copy(), self.USER_MESSAGE_TEMPLATE(prompt)]
+            if self.SYSTEM_MESSAGE is not None
+            else [self.USER_MESSAGE_TEMPLATE(prompt)]
+        )
+
+    def apply_chat_template(self, prompt: str) -> str:
+        messages = self.get_messages(prompt)
```
Comment (lapp0) [inline outlines/models/transformers.py:130]: We should probably have this be part of `class Tokenizer` and raise `NotImplementedError` by default.

**Review by lapp0 — CHANGES_REQUESTED** [review]
Thanks so much for implementing this!

There's some things I'd like to see before this is ready for merge, could you please address the following
- Test cases which ensure application of the chat template works properly for a standard case, for a missing chat template
- Consider that some users of Outlines are relying on the current behavior and make `apply_chat_template=False` the default. Perhaps we can warn the user as well if they're not setting the parameter.
- Add `apply_chat_template` explanation to the docs
- Use `apply_chat_template=True` in all `models.transformers` generator examples

**Comment by rlouf** [comment]
I don't think this is the correct design; we want to avoid kwargs as much as we possibly can. Here I would simply add a `apply_chat_template` to the model instance.

**Comment by leloykun** [comment]
@lapp0 @rlouf

How do you guys think should the interface look like?

---

Here, I mirrored Huggingface's Pipeline interface where we can specify configs/args in the constructor and (optionally) override them in the model call. I like it cuz it's more flexible. But yah, it does make things a bit more complicated and less pythonic.

I did a quick look at other libraries, and it seems that they either (1) don't auto-apply the chat templates at all or (2) have a separate code path for base & instruct/chat models (e.g. TGI & MixEval). I think there are two reasons why:

1) *It's hard to know which models are base models and which are instruct/chat models*. I thought checking whether `chat_template` is `None` or not would suffice. But some chat models apparently just leave them out (especially the older ones & third-party finetunes). Additionally, `transformers`' `PreTrainedTokenizerBase` base class has a `default_chat_template` property--so, if I'm not mistaken, we can run `tokenizer.apply_chat_template` on _all_ tokenizers without erroring out. And

2) *Some models don't support system prompts and it's hard to know which ones do and which ones don't*.

So yah, for now, we need a way to somehow let the library know that whether we're dealing with a base model or an instruct/chat model. Worst case is we might also need to ask the user to specify the system prompt. But if we're gonna force them to go all that trouble anyway, we might as well not do this by default.

I think a good compromise is to (1) not apply the chat templates by default but (2) warn the user if the chat template is specified in the tokenizer config but is not being used.

**Comment by rlouf** [comment]
```python
answer = generator(
   model.apply_chat_template(prompt)
)
```

**Comment by alonsosilvaallende** [comment]
I like to give this pull request that adds the possibility to add the `apply_chat_template`, but I wonder if it's a good idea to make it the default behavior. I have had very bad experiences with `apply_chat_template` where it add spaces or remove spaces when it shouldn't or even worse in function calling cases where it completely ignores the functions given without even raising errors (see [this example](https://gist.github.com/alonsosilvaallende/b252f6db38039158101b719bacc1bf25)). Many people might complain about something which is outside the control of Outlines.

**Comment by rlouf** [comment]
Indeed, we are not going to make it the default behavior. Users should be able to inspect the modified prompt before they pass it to the generator.

**Comment by rlouf** [comment]
@leloykun are you still working on this?

**Comment by leloykun** [comment]
@rlouf yup! I just deprioritized this in favor of other stuff; I'll get back to this soon

btw, thanks for the feedback, @alonsosilvaallende!

**Comment by Bit0r** [comment]
So can this feature still be added? A chat template is really needed in many applications.

**Comment by lapp0** [comment]
> So can this feature still be added? A chat template is really needed in many applications.

Yes, if you'd like to introduce it in a PR, I'll review and support its inclusion provided it follows the behavior discussed in this thread! :)

**Comment by RobinPicard** [comment]
Closing the PR as Outlines now has a `Chat` model input object that serves this purpose.
